File: r2dbc-prac/src/main/java/kr/jay/r2dbcprac/controller/UserController.java
package kr.jay.r2dbcprac.controller;

import org.springframework.http.HttpStatus;
import org.springframework.security.core.context.ReactiveSecurityContextHolder;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.server.ResponseStatusException;

import kr.jay.r2dbcprac.common.User;
import kr.jay.r2dbcprac.controller.dto.ProfileImageResponse;
import kr.jay.r2dbcprac.controller.dto.SignupUserRequest;
import kr.jay.r2dbcprac.controller.dto.UserResponse;
import kr.jay.r2dbcprac.service.UserService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import reactor.core.publisher.Mono;

/**
 * UserController
 *
 * @author jaypark
 * @version 1.0.0
 * @since 2023/07/17
 */

@Slf4j
@RequiredArgsConstructor
@RestController
@RequestMapping("/api/users")
public class UserController {

	private final UserService userService;

	@GetMapping("/{userId}")
	public Mono<UserResponse> getUserById(@PathVariable String userId) {
		log.info("=====================");
		return ReactiveSecurityContextHolder.getContext()
			.flatMap(context->{
				final String userName = context.getAuthentication().getName();

				if (!userId.equals(userName))
					return Mono.error(new ResponseStatusException(HttpStatus.UNAUTHORIZED));

				return userService.findById(userId)
					.map(this::map);
			}).switchIfEmpty(Mono.error(new ResponseStatusException(HttpStatus.NOT_FOUND)));
	}

	@ResponseStatus(HttpStatus.CREATED)
	@PostMapping("/signup")
	Mono<UserResponse> signupUser(
		@RequestBody final SignupUserRequest request
	) {
		return userService.createUser(
			request.getName(),
			request.getAge(),
			request.getPassword(),
			request.getProfileImage()
		).map(this::map);

	}


	private UserResponse map(final User user) {
		return new UserResponse(
			user.getId(),
			user.getName(),
			user.getAge(),
			user.getFollowCount(),
			user.getProfileImage()
				.map(image ->
					new ProfileImageResponse(image.getId(), image.getName(), image.getUrl())
				)
		);
	}

}

File: spring-coroutine-prac/src/main/kotlin/kr/jay/springcoroutineprac/controller/UserController.kt
package kr.jay.springcoroutineprac.controller

import kr.jay.r2dbcprac.common.Image
import kr.jay.r2dbcprac.common.User
import kr.jay.r2dbcprac.controller.dto.ProfileImageResponse
import kr.jay.r2dbcprac.controller.dto.SignupUserRequest
import kr.jay.r2dbcprac.controller.dto.UserResponse
import kr.jay.springcoroutineprac.service.UserCoroutineService
import org.springframework.http.HttpStatus
import org.springframework.security.core.context.ReactiveSecurityContextHolder
import org.springframework.security.core.context.SecurityContext
import org.springframework.web.bind.annotation.*
import org.springframework.web.server.ResponseStatusException
import reactor.core.publisher.Mono

/**
 * UserCoroutineController
 *
 * @author jaypark
 * @version 1.0.0
 * @since 2023/08/24
 */

@RequestMapping("/api/users")
@RestController
class UserController(
    private val userService: UserCoroutineService
) {

    @GetMapping("/{userId}")
    fun getUserById(
        @PathVariable userId: String
    ): Mono<UserResponse> {
        return ReactiveSecurityContextHolder.getContext()
            .flatMap { context: SecurityContext ->
                val userName: String = context.authentication.name
                if (userId != userName) return@flatMap Mono.error<UserResponse>(
                    ResponseStatusException(
                        HttpStatus.UNAUTHORIZED
                    )
                )
                userService.findById(userId)
                    .map(this::map)
                    .switchIfEmpty(Mono.error(ResponseStatusException(HttpStatus.NOT_FOUND)))
            }
    }

    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping("/signup")
    fun signupUser(
        @RequestBody request: SignupUserRequest
    ): Mono<UserResponse> {
        return userService.createUser(
            request.name,
            request.age,
            request.password,
            request.profileImage
        ).map(this::map)
    }


    private fun map(user: User): UserResponse {
        return UserResponse(
            user.id,
            user.name,
            user.age,
            user.followCount,
            user.profileImage
                .map { image: Image ->
                    ProfileImageResponse(
                        image.id,
                        image.name,
                        image.url
                    )
                }
        )
    }
}